pkg/policy/cell/policy_importer.go

```diff
@@ -0,0 +1,395 @@
+}
+
+func newPolicyImporter(cfg policyImporterParams) PolicyImporter {
+	i := &policyImporter{
+		log:          cfg.Log,
+		repo:         cfg.Repo,
+		epm:          cfg.EndpointManager,
+		ipc:          cfg.IPCache,
+		monitorAgent: cfg.MonitorAgent,
+
+		q: make(chan *policytypes.PolicyUpdate, option.Config.PolicyQueueSize),
+
+		prefixesByResource: map[ipcachetypes.ResourceID][]netip.Prefix{},
+	}
+
+	buf := stream.Buffer(
+		stream.FromChannel(i.q),
+		option.Config.PolicyQueueSize, 10*time.Millisecond,
+		concat)
+
+	cfg.JobGroup.Add(job.Observer("policy-importer", i.processUpdates, buf))
+
+	return i
+}
+
+// ResourceIDAnonymous is the anonymous ipcache resource used as a placeholder
+// for policies that allocate CIDRs but do not have an owning resource.
+// (This is only used for policies created by the local API).
+const ResourceIDAnonymous = "policy/anonymous"
+
+func (i *policyImporter) UpdatePolicy(u *policytypes.PolicyUpdate) {
+	i.q <- u
+}
+
+func concat(buf []*policytypes.PolicyUpdate, in *policytypes.PolicyUpdate) []*policytypes.PolicyUpdate {
+	buf = append(buf, in)
+	return buf
+}
+
+// updatePrefixes determines the set of prefixes "owned" by a given resource and applies them
+// to the ipcache.
+// Write lock must be held.
+//
+// If the ipcache has started, it waits up to a configurable deadline for the prefixes
+// to be allocated.
+//
+// It returns the set of stale prefixes that should be deallocated after policy updates are complete.
+func (i *policyImporter) updatePrefixes(ctx context.Context, updates []*policytypes.PolicyUpdate) (toPrune map[ipcachetypes.ResourceID][]netip.Prefix) {
+	if i.ipc == nil {
+		return
+	}
+
+	// The set of all prefixes that belong to a resource.
+	toAllocate := map[ipcachetypes.ResourceID][]netip.Prefix{}
+	prefixSource := map[ipcachetypes.ResourceID]source.Source{}
+	toPrune = map[ipcachetypes.ResourceID][]netip.Prefix{}
+
+	// First, gather all new prefixes and index them by resource.
+	//
+	// For rules without an owning resource (k8s policy or file on disk),
+	// we never de-allocate prefixes. Otherwise, we will track prefixes
+	// by resource and de-allocate unused prefixes after policy is applied.
+	for _, upd := range updates {
+		prefixes := policy.GetCIDRPrefixes(upd.Rules)
+		if upd.Resource == "" {
+			// edge-case: no owning resource.
+			// Allocate prefixes with a placeholder.
+			if len(prefixes) == 0 {
+				continue
+			}
+			// since anonymous prefixes may come from multiple sources,
+			// we append to the list
+			toAllocate[ResourceIDAnonymous] = append(toAllocate[ResourceIDAnonymous], prefixes...)
+			prefixSource[ResourceIDAnonymous] = upd.Source // This could overwrite if there are multiple sources, but in practice there aren't
+
+		} else {
+			// Standard case: there is an owning resource.
+			// Track the complete set of per-prefix resources.
+			// We want empty sets here!
+			toAllocate[upd.Resource] = prefixes
```

Comment:
Good question. For anonymous policies (those created by the local REST api), we don't clean up unused prefixes since we have no easy way to track their ownership. For policies with an owning reference (all k8s objects, as well as files on disk), we need to track empty here so we later deallocate any prefixes which previously belonged to the resource.